ansi_c_parsert: construct with message handler

tautschnig opened the pull request:
This both avoids an object of static lifetime as well as it fixes the (transitive) use of the deprecated messaget() constructor.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
C and C++ parsers: construct with message handler

This both avoids an object of static lifetime as well as it fixes the
(transitive) use of the deprecated messaget() constructor. The C scanner
is now fully reentrant.

feliperodri reviewed (approved):
All changes in the contracts wrangler LGTM.

peterschrammel commented on regression/cbmc/fault_localization-all_properties2/test.desc:
```diff
@@ -1,4 +1,4 @@
-KNOWNBUG
+FUTURE
```
❓ This change seems unrelated? Maybe put into a separate commit.

tautschnig replied:
Done.

peterschrammel commented on src/ansi-c/ansi_c_parser.h:
```diff
@@ -35,14 +34,14 @@ class ansi_c_parsert:public parsert
+  bool parse() override;
 
-  virtual void clear() override
+  void clear() override
```
❓ Is this function still used anywhere?

tautschnig replied:
I will clean this up across the type hierarchy (removing just this override seems dangerous, as any use of `.clear()` would just invoke `parsert::clear()`, which would end up in incomplete cleanup) in a follow-up PR.

tautschnig committed:
Mark flaky fault-localization test as FUTURE

The test will sometimes pass, depending on what model the solver
produces. When such a passing model happens to be is produced, it breaks
our checking of "KNOWNBUG" tests.
